PR for release 2.2.0 "Ulm" (#573)
nf-core/eager: docs/usage.md

This module utilises `samtools view` and `filter_bam_fragment_length.py`

#### `--run_bam_filtering`

Turns on the bam filtering module for either mapping quality filtering or
unmapped read treatment.

> :warning: this is **required** for metagenomic screening!

#### `--bam_mapping_quality_threshold`

Specify a mapping quality threshold for mapped reads to be kept for downstream
analysis. By default keeps all reads and is therefore set to `0` (basically
doesn't filter anything).

> Modifies samtools view parameter: `-q`

#### `--bam_filter_minreadlength`

Specify minimum length of mapped reads. This filtering will apply at the same
time as mapping quality filtering.

If used _instead_ of minimum length read filtering at AdapterRemoval, this can
be useful to get more realistic endogenous DNA percentages, when most of your
reads are very short (e.g. in single-stranded libraries) and would otherwise be
discarded by AdapterRemoval (thus making an artificially small denominator for a
typical endogenous DNA calculation). Note in this context you should not perform
mapping quality filtering nor discarding of unmapped reads to ensure a correct

> 'denominator' of 'all reads', for the Endogenous DNA calculation.

Reviewer: ```suggestion
denominator of all reads, for the endogenous DNA calculation.
```

> Modifies filter_bam_fragment_length.py parameter: `-l`

#### `--bam_unmapped_type`

Defines how to proceed with unmapped reads: `'discard'` removes all unmapped
reads, `keep` keeps both unmapped and mapped reads in the same BAM file, `'bam'`
keeps unmapped reads as BAM file, `'fastq'` keeps unmapped reads as FastQ file,
`both` keeps both BAM and FASTQ files. Default is `discard`.  `keep` is what
would happen if `--run_bam_filtering` was _not_ supplied.

Note that in all cases, if `--bam_mapping_quality_threshold` is also supplied,
mapping quality filtering will still occur on the mapped reads.

:warning: `--bam_unmapped_type 'fastq'` is **required** for metagenomic
screening!

> Modifies samtools view parameter: `-f4 -F4`

### Read DeDuplication Parameters

> If using TSV input, deduplication is performed library, i.e. after lane merging.

Reviewer: ```suggestion
If using TSV input, deduplication is performed per library, i.e. after lane merging.
```

_copied_ from a `work/` directory into the results folder. You should clean the
work directory with the command to ensure non-redundancy and large HDD
footprints!

#### `--database`

Specify the path to the _directory_ containing your taxonomic classifier's
database (malt or kraken).

For Kraken2, it can be either the path to the _directory_ or the path to the
`.tar.gz` compressed directory of the Kraken2 database.

#### `--metagenomic_min_support_reads`

Specify the minimum number of reads a given taxon is required to have to be
retained as a positive 'hit'.  
For malt, this only applies when `--malt_min_support_mode` is set to 'reads'.
Default: `1`.

> Modifies MALT or kraken_parse.py parameter: `-sup` and `-c` respectively

#### `--percent_identity`

Specify the minimum percent identity (or similarity) a sequence must have to the
reference for it to be retained. Default is `85`

Only used when `--metagenomic_tool malt` is also supplied.

> Modifies MALT parameter: `-id`

#### `--malt_mode`

Use this to run the program in 'BlastN', 'BlastP', 'BlastX' modes to align DNA
and DNA, protein and protein, or DNA reads against protein references

> respectively. respectively. Ensure your database matches the mode. Check the

Reviewer 2: ```suggestion
respectively. Ensure your database matches the mode. Check the
```

> Modifies MaltExtract parameter: `--matches`

#### `--maltextract_megansummary`

Export 'minimal' summary files (i.e. without alignments) that can be loaded into
[MEGAN6](https://doi.org/10.1371/journal.pcbi.1004957). By default turned off.

Only when `--metagenomic_tool malt` is also supplied.

> Modifies MaltExtract parameter: `--meganSummary`

#### `--maltextract_percentidentity`

Minimum percent identity alignments are required to have to be reported. Higher
values allows fewer mismatches between read and reference sequence, but
therefore will provide greater confidence in the hit. Lower values allow more
mismatches, which can account for damage and divergence of a related
strain/species to the reference. Recommended to set same as MALT parameter or
higher. Default: `85.0`.

Only when `--metagenomic_tool malt` is also supplied.

> Modifies MaltExtract parameter: `--minPI`

#### `maltextract_topalignment`

Use the best alignment of each read for every statistic, except for those
concerning read distribution and coverage. Default: off.

Only when `--metagenomic_tool malt` is also supplied.

> Modifies MaltExtract parameter: `--useTopAlignment`

### Clean up

> Once completed a run has completed, you will have _lots_ of (some very large)

Reviewer 2: ```suggestion
Once a run has completed, you will have _lots_ of (some very large)
```

> intermediate files in your output directory, within the directory named `work`.

Reviewer 2: sounds weird.

Once you have verified your run completed correctly and everything in the module
output directories are present as you expect and need, you can perform a clean

> up.

Reviewer 2: ```suggestion
output directories are present as you expect and need, you can perform a clean-up.
```

> > **Important**: Once clean up is completed, you will _not_ be able to re-rerun

Reviewer 2: ```suggestion
> **Important**: Once clean-up is completed, you will _not_ be able to re-rerun
```

should be in a valid [NCBI taxonomy](https://www.ncbi.nlm.nih.gov/taxonomy) name
format.

Only when `--metagenomic_tool malt` is also supplied.

#### `--maltextract_ncbifiles`

Path to directory containing containing the NCBI resource tree and taxonomy
table files (ncbi.tre and ncbi.map; available at the [HOPS
repository](https://github.com/rhuebler/HOPS/Resources)).

Only when `--metagenomic_tool malt` is also supplied.

#### `--maltextract_filter`

Specify which MaltExtract filter to use. This is used to specify what types of
characteristics to scan for. The default will output statistics on all
alignments, and then a second set with just reads with one C to T mismatch in
the first 5 bases. Further details on other parameters can be seen in the [HOPS
documentation](https://github.com/rhuebler/HOPS/#maltextract-parameters).
Options: `'def_anc'`, `'ancient'`, `'default'`, `'crawl'`, `'scan'`, `'srna'`,
'assignment'. Default: `'def_anc'`.

Only when `--metagenomic_tool malt` is also supplied.

> Modifies MaltExtract parameter: `-f`

#### `--maltextract_toppercent`

Specify percent of top alignments for each read to be considered for each node.

> Default: `0.01`.

Reviewer 3: Ugh, no it uses a float notation 1 = 100%, so 0.01 is 1%, no idea why the developer changed the way to record it from MALT.

> Default: `0.01`.

Reviewer: Then you should change the phrasing to float. Because later on you have another percentage parameter, where it actually is in percentages.

> Default: `0.01`.

Reviewer 3: Yup, done:

```
Specify frequency of top alignments for each read to be considered for each node.
Default is 0.01, i.e. 1% of all reads (where 1 would correspond to 100%).

> :warning: this parameter follows the same concept as `--malt_top_percent` but
> uses a different notation i.e. integer (MALT) versus float (MALTExtract)

Default: `0.01`.

Only when `--metagenomic_tool malt` is also supplied.

> Modifies MaltExtract parameter: `-a`
```

Only when `--metagenomic_tool malt` is also supplied.

> Modifies MaltExtract parameter: `-a`

> #### `--maltextract_destackingoff`

Reviewer 3: :+1:

```

However, in very rare cases, minor fixes to a version will be pushed out without
a version number bump. This can confuse nextflow slightly, as it thinks you
already have the 'broken' version from your original pipeline download.

If when running the pipeline you don't see any changes in the fixed version when
running it, you can try removing your nextflow EAGER cache typically stored in
your home directory with

```bash
rm -r ~/.nextflow/assets/nf-core/eager
```

And re-pull the pipeline with the command above. This will install a fresh
version of the version with the fixes.

### Input files not found

If no file, only one input file, or only 'read one' and not 'read two' is picked
up then something is likely wrong with your input file declaration
([`--input`](#--input)):

1. The path must be enclosed in quotes (`'` or `"`)
2. The path must have at least one `*` wildcard character. This is even if you
   are only running one paired end sample.
3. When using the pipeline with paired end data, the path must use `{1,2}` or
   `{R1,R2}` notation to specify read pairs.
4. If you are running single-end data make sure to specify `--single_end`

> **Important**: The pipeline can't take a list of multiple input files - it takes

Reviewer 3: Nice catch!

nextflow clean -f -k
```

This will make your system administrator very happy as you will _halve_ the
hard drive footprint of the run, so be sure to do this!

## Troubleshooting and FAQs

### My pipeline update doesn't seem to do anything

To download a new version of a pipeline, you can use the following, replacing
`<VERSION>` the the corresponding version.

```bash
nextflow pull nf-core/eager -r <VERSION>
```

However, in very rare cases, minor fixes to a version will be pushed out without
a version number bump. This can confuse nextflow slightly, as it thinks you
already have the 'broken' version from your original pipeline download.

If when running the pipeline you don't see any changes in the fixed version when
running it, you can try removing your nextflow EAGER cache typically stored in
your home directory with

```bash
rm -r ~/.nextflow/assets/nf-core/eager
```

And re-pull the pipeline with the command above. This will install a fresh
version of the version with the fixes.

> ### Warning about sticked on revision
> 
> If you get a warning like the following:
> 
> ```bash
> Project nf-core/eager currently is sticked on revision: dev -- you need to specify explicitly a revision with the option `-r` to use it
> ```
> 
> This is a nextflow error. This means that you have multiple versions of
> nf-core/eager downloaded (e.g. 2.0.0, 2.1.0, 2.1.1, dev etc.). With every
> `nextflow run nf-core/eager` command you must always specify which one you want
> to use with `-r` - there is no 'default' version it will use.
> 
> For example you must run :
> 
> ```bash
> nextflow run nf-core/eager -r 2.1.0 --reads '/<path>/<to>/data/*_{R1,R2}_*.fq.gz' <...>
> ```
> 
> Specifying the version of the run you are using is highly recommended, as it
> helps in full reproducibility. Whereby if you record the whole command _with_
> the version for your publication or internal reports, then anyone who wants to
> check your work can use the exact version you also used (including all internal
> tools).
> 
> You can see more information on the nextflow documentation
> [here](https://www.nextflow.io/docs/latest/sharing.html?highlight=revision#handling-revisions).

Reviewer 4: ```suggestion
```

Already there: https://nf-co.re/usage/troubleshooting#warning-about-sticked-on-revision

On the other hand encapsulating the path in quotes will allow Nextflow to
evaluate the paths.

```bash
nextflow run nf-core/eager --input "/path/to/sample_*/*.fq.gz"
```

### The pipeline crashes almost immediately with an early pipeline step

Sometimes a newly downloaded and set up nf-core/eager pipeline will encounter an
issue where a run almost immediately crashes (e.g. at `fastqc`,
`output_documentation` etc.) saying the tool could not be found or similar.

#### I am running Docker

You may have an outdated container. This happens more often when running on the
`dev` branch of nf-core/eager, because docker will _not_ update the container on
each new commit, and thus may not get new tools called within the pipeline code.

To fix, just re-pull the nf-core/eager docker container manually with:

```bash
docker pull nfcore/eager:dev
```

#### I am running Singularity

If you're running singularity, it could be that nextflow cannot access your
singularity image properly - often due to missing bind paths.

See
[here](https://nf-co.re/usage/troubleshooting#cannot-find-input-files-when-using-singularity)
for more information.

> ### The pipeline has crashed with an error but Nextflow is still running

Reviewer 2: Seems like you have the same problem...

Reviewer 3: What do you mean here?

Reviewer 2: Ah, I think you can ignore that. @apeltzer did a review on our proteomicslfq and I think we discussed that problem as well. I made a "Known issue" out of it as well.

Author: ❤️  that people are finding these and making overall progress on reporting this back to the main nf-core docs :-)

Caused by:
  Process `library_merge` input file name collision -- There are multiple input files for each of the following file names: JK2782.mapped_rmdup.bam.csi, JK2782.mapped_rmdup.bam
Tip: you can try to figure out what's wrong by changing to the process work dir and showing the script file named `.command.sh`
Execution cancelled -- Finishing pending tasks before exit
```

In this case: for lane merging errors, you can give 'fake' lane IDs to ensure
they are unique (e.g. if one library was sequenced on Lane 8 of two HiSeq runs,
specify lanes as 8 and 16 for each FASTQ file respectively). For library merging
errors, you must modify your `Library_ID`s accordingly, to make them unique.

### I specified a module and it didn't produce the expected output

Possible options:

1. Check there if you have a typo in the parameter name. Nextflow _does not_
   check for this
2. Check that an upstream module was turned on (if a module requires the output
   of a previous module, it will not be activated unless it receives the output)

### I get a unable to acquire lock

Errors like the following

```bash
Unable to acquire lock on session with ID 84333844-66e3-4846-a664-b446d070f775
```

> Normally suggest a previous nextflow run (on the same folder) was not cleanly

Reviewer 2: ```suggestion
normally suggest a previous nextflow run (on the same folder) was not cleanly
```
